ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs

```diff
@@ -34,7 +34,37 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for DummyInstruction<E
     fn construct_circuit(
         circuit_builder: &mut CircuitBuilder<E>,
     ) -> Result<Self::InstructionConfig, ZKVMError> {
-        DummyConfig::construct_circuit(circuit_builder, I::INST_KIND.codes())
+        let codes = I::INST_KIND.codes();
+
+        // ECALL can do everything.
+        let is_ecall = matches!(codes.kind, InsnKind::EANY);
+
+        // Regular instructions do what is implied by their format.
+        let (with_rs1, with_rs2, with_rd) = match codes.format {
+            _ if is_ecall => (true, true, true),
+            InsnFormat::R => (true, true, true),
+            InsnFormat::I => (true, false, true),
+            InsnFormat::S => (true, true, false),
+            InsnFormat::B => (true, true, false),
+            InsnFormat::U => (false, false, true),
+            InsnFormat::J => (false, false, true),
+        };
+        let with_mem_write = matches!(codes.category, InsnCategory::Store) || is_ecall;
+        let with_mem_read = matches!(codes.category, InsnCategory::Load);
+        let branching = matches!(codes.category, InsnCategory::Branch)
+            || matches!(codes.kind, InsnKind::JAL | InsnKind::JALR)
+            || is_ecall;
```

Comment:
Most of `Ecall` instructions follow the default branching rule `pc' = pc + 4` except `Ecall/Halt`.

Reply:
That’s ok, who can do more, can do less. I do imagine that is how I will do the complex ecalls: jump to handler code.